### ISSUE TITLE
[Bug] Diag() on MatmulLazyTensor using RootLazyTensors is expensive

### ISSUE THREAD

**Issue by wjmaddox** [cornellius-gp/gpytorch#847]
# 🐛 Bug

## To reproduce

** Code snippet to reproduce **
```python
a = torch.randn(1761, 1761)
b = torch.randn(1761, 1761)

a_root = gpytorch.lazy.RootLazyTensor(a)
mm2 = gpytorch.lazy.MatmulLazyTensor(a_root, gpytorch.lazify(b))

start = time.time()
mm2.diag()
end = time.time() - start
# This takes 84.2 seconds.

((a @ a.t()) @ b).diag()
# this takes 0.2 seconds.
```
By comparison, using matrices explicitly takes about 0.2 seconds on a CPU.

** Stack trace/error message **

I've done my best to track this down as .diag() seems to call the MatmulLazyTensor _get_indices_ method, which expands out the matrices (see [here](https://github.com/cornellius-gp/gpytorch/blob/46aeefdff0438c50419eba39ca5a774a8872d174/gpytorch/lazy/matmul_lazy_tensor.py#L42) ).
The call then goes into RootLazyTensors _get_indices method which expands out the sizing to be quite large - e.g. right_tensor has shape `torch.Size([1, 1761, 1761])` and left_tensor has shape `torch.Size([1761, 1, 1761])`. 

[See here](https://github.com/cornellius-gp/gpytorch/blob/46aeefdff0438c50419eba39ca5a774a8872d174/gpytorch/lazy/root_lazy_tensor.py#L34)

## Expected Behavior

I think MatmulLazyTensor should evaluate itself with matmuls against `torch.eye` but it's not clear if that would be faster.

## System information

**Please complete the following information:**
- gpytorch 0.3.5
- torch 1.1.0
- Macbook Pro with Intel i5 processor

## Additional context
See above.

**Comment by wjmaddox** [comment]
Attempting to decompose the matrix multiplication as 

`mm3 = gpytorch.lazy.MatmulLazyTensor(a_root.root, 
                                     gpytorch.lazy.MatmulLazyTensor(a_root.root.t(), mm2.right_lazy_tensor))
mm3.diag()`

seems to be similarly computationally intensive suggesting it could be a problem with indexing?

**Comment by jacobrgardner** [comment]
The problem here is that it's using the default `diag` behavior when the tensors in a MatmulLazyTensor aren't non lazy, because in general evaluating can make things slower.

With that said, you are creating a full rank RootLazyTensor, for which operations are unlikely to be faster than just creating a NonLazyTensor out of `a @ a.t()` -- is there a reason you are doing this? In that case I'd probably recommend making a `NonLazyTensor` out of `a @ a.t()` and `b`, and then making a `MatmulLazyTensor` out of those two.

**Comment by wjmaddox** [comment]
I noticed the issue in computing predictive variances in Bayesian linear regressions (which expand into parameter space), where it seems like the root lazy tensor part is a distraction.

The actual code is pretty much like:

```python
# not actually a nonlazy tensor but for demonstration purposes
b = gpytorch.lazify(torch.randn(1000, 1761))`

# get something that looks like I - SS'
a_root = gpytorch.lazy.RootLazyTensor(torch.randn(1761, 90))
a_sub_part = (-1 * a_root).add_jitter(1.)

# predictive covariance is b (I - ss') b'
covar = b @ (a_sub_part @ b.t())

# get diagonal of the matrix
covar.diag()
```

**Comment by jacobrgardner** [comment]
If you want to do solves with `S'S + I`, where `S` is `n x d`, you can do `(S'S + I)^{-1} = I - S'(I_{n} + SS')^{-1}S`, which lets you solve an `n x n` system instead of a `d x d` system. You can use identities like this for both the predictive mean and variance.

If both `n` and `d` are very large and roughly equal, then you are just generally in trouble / unlikely to get much faster speed out of linear regression than any other GP without using variational inference or other methods for stochastic optimization.

**Comment by wjmaddox** [comment]
Yes, I know that. In this particular situation, although `d >> n`, it is the case that `SS'v` is faster than `S'Sv`. Here `S` corresponds to the Jacobian matrix of a neural net.

A less memory intensive solution is to split the covariance matrix:
```python

covar2 = gpytorch.lazy.RootLazyTensor(b) + -1 * gpytorch.lazy.RootLazyTensor(b @ a_root.root)
covar2.diag()

```
